### PR TITLE
Alternative approach to the layout outer padding

### PR DESCRIPTION
**Pull request by youknowriad**
Alternative to #35919 

Instead of using a dedicated `outerPadding` property on the layout object. This PR uses `padding`, so for instance if a user sets a padding to a group block and also add the same time defines content/wide widths, the padding applied to the block is used to compute the negative margins.

That said, when the layout is defined as "inherit: true", the padding control in the container block is hidden and the "padding" defined in the theme.json file (settings.layout.padding) is used instead.

What do you think? It feels a bit better than #35919 because there's no padding conflicts in the UI.

**Cons:**

- If a user defines a padding on a group block and then after that triggers the "inherit layout" checkbox, the initially defined padding is going to be ignored ant the control hidden but in order for the styles to apply correctly the padding applied using the layout code has `!important`
- The other con is that padding and layout are kind of interconnected with this PR.

### CODE REVIEW

**Comment by kjellr** [comment]
@youknowriad: @jffng and I are both getting editor errors when we try this PR: 

Screen Shot 2021-11-04 at 11 07 47 AM

Error: 
```
save@https://gutenberg.test/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=876a1b8ffced55f15985649073f17ad5:3:17985
renderWithHooks@https://unpkg.com/react-dom@17.0.1/umd/react-dom.development.js:15015:29
updateFunctionComponent@https://unpkg.com/react-dom@17.0.1/umd/react-dom.development.js:17386:37
callCallback@https://unpkg.com/react-dom@17.0.1/umd/react-dom.development.js:3942:21
dispatchEvent@[native code]
invokeGuardedCallbackDev@https://unpkg.com/react-dom@17.0.1/umd/react-dom.development.js:3991:31
invokeGuardedCallback@https://unpkg.com/react-dom@17.0.1/umd/react-dom.development.js:4053:38
beginWork$1@https://unpkg.com/react-dom@17.0.1/umd/react-dom.development.js:23994:30
performUnitOfWork@https://unpkg.com/react-dom@17.0.1/umd/react-dom.development.js:22809:25
workLoopSync@https://unpkg.com/react-dom@17.0.1/umd/react-dom.development.js:22737:24
renderRootSync@https://unpkg.com/react-dom@17.0.1/umd/react-dom.development.js:22700:21
performSyncWorkOnRoot@https://unpkg.com/react-dom@17.0.1/umd/react-dom.development.js:22323:34
performSyncWorkOnRoot@[native code]
@https://unpkg.com/react-dom@17.0.1/umd/react-dom.development.js:11357:36
unstable_runWithPriority@https://unpkg.com/react@17.0.1/umd/react.development.js:2942:26
flushSyncCallbackQueueImpl@https://unpkg.com/react-dom@17.0.1/umd/react-dom.development.js:11352:28
flushSyncCallbackQueue@https://unpkg.com/react-dom@17.0.1/umd/react-dom.development.js:11339:31
scheduleUpdateOnFiber@https://unpkg.com/react-dom@17.0.1/umd/react-dom.development.js:21923:33
dispatchAction@https://unpkg.com/react-dom@17.0.1/umd/react-dom.development.js:16169:28
dispatchAction@[native code]
@https://gutenberg.test/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=876a1b8ffced55f15985649073f17ad5:52:11374
```

File: packages/block-editor/src/layouts/flow.js [hunk 1]
```diff
@@ -125,6 +133,8 @@ export default {
 
 					${ appendSelectors( selector, '> [data-align="full"]' ) } {
 						max-width: none;
+						margin-left: calc( -1 * ${ padding.left || 0 } ) !important;
```
Comment (jffng) [inline packages/block-editor/src/layouts/flow.js]: I think these also should be optionally chained? 
```suggestion
						margin-left: calc( -1 * ${ padding?.left || 0 } ) !important;
```

**Comment by youknowriad** [comment]
yep, should be fixed now. The padding defined in the layout should be something like: `"padding": { "left": "var(--wp--custom--spacing--small, 1.25rem)", "right": "var(--wp--custom--spacing--small, 1.25rem)" }
`

**Comment by kjellr** [comment]
Alright thanks — I can access the editor again. I drafted up a quick PR in https://github.com/WordPress/twentytwentytwo/pull/209 to test this against. 

<strike>I'm having trouble understanding how to actually make things go full width though. I'm not seeing negative margins applied anywhere? It sounded like I should be able to (for example) put this in the root of my site editor, and have the image go full-width, right?</strike>

```
<!-- wp:group {"layout":{"inherit":true}} -->
<div class="wp-block-group"><!-- wp:image {"align":"full","sizeSlug":"full"} -->
<figure class="wp-block-image alignfull size-full"><img src="https://cldup.com/NCVsLRzaSd.png" alt=""/></figure>
<!-- /wp:image --></div>
<!-- /wp:group -->
```

EDIT: Nevermind, the suggestion in https://github.com/WordPress/twentytwentytwo/pull/209#discussion_r742943717 got this working!

**Comment by kjellr** [comment]
This works pretty well! The only major hesitation I have is this part: 

> That said, when the layout is defined as "inherit: true", the padding control in the container block is hidden and the "padding" defined in the theme.json file (settings.layout.padding) is used instead.

Here's a visual representation of this: 

Screen Shot 2021-11-04 at 12 03 00 PM

The theme has left/right padding and no top padding, so that's what's applied to the block when inherit layout is true. There's no way to modify that padding at all. I feel like we should be able to still access padding controls, and if we edit those, the negative margins should just shut off automatically. Is that possible?

**Comment by youknowriad** [comment]
> The theme has left/right padding and no top padding, so that's what's applied to the block when inherit layout is true. There's no way to modify that padding at all. I feel like we should be able to still access padding controls, and if we edit those, the negative margins should just shut off automatically. Is that possible?

I'l reluctant because this is not how it works with the layout styles (widths). When you inherit, you ignore values defined in the block itself. So if we change it for padding, I wonder if we should be consistent.

So: if block specific value is defined use it, otherwise use the inherited value. Also, the question here is do you think we should prefill the padding control with the layout inherited value? Also, how do you clear the value? Right now clearing a value means making it undefined which would revert to the inherited value, it could become confusing.

**Comment by kjellr** [comment]
Yeah, that makes sense. This is a complicated problem. 😅 

I'm thinking that we're probably not ready to merge this for the feature freeze — it doesn't feel quite ready. Since it's really just an iteration on existing functionality we can probably keep thinking through it through Beta anyway? I'm not sure.

**Comment by youknowriad** [comment]
I agree that there's no need to rush it 👍 .

**Comment by richtabor** [comment]
>  I feel like we should be able to still access padding controls, and if we edit those, the negative margins should just shut off automatically. Is that possible?

Whew, this is a complicated one. So you're thinking that if the padding control is modified, then the negative margins wouldn't work any longer?

Other question, this doesn't solve the fullwidth/negative margins for any other alignfull blocks, correct?

**Comment by kjellr** [comment]
> So you're thinking that if the padding control is modified, then the negative margins wouldn't work any longer?

That's right. My general thinking is that if you've added padding to a specific block this way, you're overriding the defaults and choosing to have that padding surround every interior block. That generally works for my mental model, but I'm not sure how universal it is. Like you said, this is a complicated issue!

> Other question, this doesn't solve the fullwidth/negative margins for any other alignfull blocks, correct?

The PR currently works for any alignfull block whose direct parent inherits the default layout.

**Comment by kjellr** [comment]
Here's an example of where I found the automatically-added padding confusing: In the post/page editor, I have a Group block that's wide-aligned. I've activated "Inherit default layout" because I'd like this block to contain one wide-width block, and one standard-width text block. Here's how that looks today: 

Screen Shot 2021-11-23 at 7 54 50 AM

This is as I'd expect it to be. The Group block parent takes up the full "wide" width of the page, and so does the wide-width child. The standard-aligned block is smaller than that. The site padding exists, but it's on the left and right of the entire block (as expected). 

But with this PR active, the Group block parent gets extra padding, so it displays like this: 

Screen Shot 2021-11-23 at 7 48 14 AM

There's no need to double-up on the padding here. The "site padding" is already present outside of this block. And there's no way for me to remove it currently.

**Comment by youknowriad** [comment]
> There's no need to double-up on the padding here. The "site padding" is already present outside of this block. And there's no way for me to remove it currently.

For me the issue here is that you should have never used "inherit layout" on that group block, instead you should have defined a custom layout (custom widths), the reason is simple: the page canvas is full width, meaning a layout that is suitable there (the default layout) is not something you should apply to a "wide" block because it breaks there (regardless of padding).

Example: the theme defines: `1000px`/`800px` as widths for the default layout, it doesn't really make sense to use the same layout for a block that is maximum `1000px` for me (wide group), because a wide and full width will have similar behavior there. 

In other words for me: `inherit layout` only makes sense for blocks who's width is the full window size. these can be nested or top level (which makes this rule hard to enforce) but applying "inherit" is not suitable in any other situation for me.

**Comment by kjellr** [comment]
> In other words for me: inherit layout only makes sense for blocks who's width is the full window size. these can be nested or top level (which makes this rule hard to enforce) but applying "inherit" is not suitable in any other situation for me.

What's the benefit to us having this control available on blocks that are not full-width then? Can/should we filter that out? The majority of blocks the user sees these controls on do not fit that criteria.

**Comment by youknowriad** [comment]
@kjellr Think of it as a preset for theme authors, so they don't have to redefine the same layout over and over across blocks and templates.

I'd love to find a way to hide it in the situations where it makes less sense but it's very hard to know statically. For instance it makes sense in a query loop inside a group inside a group inside a group inside a template part if the top level is full width.

**Comment by kjellr** [comment]
Where are we at with this one? I think it's probably close, but I still think having the option to clear these paddings for child blocks is important. If that's something we'd consider trying to add, do we need Design to help figure out answers for [these questions](https://github.com/WordPress/gutenberg/pull/36214#issuecomment-961776596)? 

> Also, the question here is do you think we should prefill the padding control with the layout inherited value? Also, how do you clear the value? Right now clearing a value means making it undefined which would revert to the inherited value, it could become confusing.

**Comment by youknowriad** [comment]
I'm personally not confident about showing the padding control or allowing overriding the "default inherited one" because it creates a precedent without how the inherited layout properties work.

 - Why not keep the wide/full widths either?
 - Why not prefill these with the values from theme.json?
 - Why don't we prefill the values for other styles either, think all of the fontsizes, colors... where the default values are defined in theme.json

I don't think it's something to be done specifically for this PR but more something to be thought of holistically and not something we can do properly for 5.9.

**Comment by youknowriad** [comment]
I'm personally happy to ship this as is if folks thinks it's a decent compromise. I can't think of a better one for now that won't have bigger impacts.

**Comment by kjellr** [comment]
I'm still pretty uncomfortable with because it takes an already complex control (Layout) and adds additional complicated behavior to it. Here are a few of my concerns: 

---

1. In general, it's very confusing for a user to lose all their custom paddings when they flip the layout toggle: 

spacings

We could adjust the label to say "Inherit default layout and padding", but the layout control is already pretty opaque to the user (cc https://github.com/WordPress/gutenberg/pull/36675), and I'm not sure if that helps or adds complexity. 

---

2. In general, I think there's a conceptual difference between the "site" padding and the padding that I'd want to add to individual layout-enabled blocks on the site. It's quite likely that many themes will want to have their top padding set to `0` to allow header images/colors to reach the top of the screen. But I don't think the same applies to layout-enabled blocks elsewhere in the site. They're totally different contexts. 

---

3. In practice, this PR makes a common layout like this Twenty Twenty-Two pattern more difficult to achieve: 

Screen Shot 2021-12-13 at 1 34 48 PM

This content in the middle of this block is wide-aligned, and sits inside of a full-width gray Group block. The padding on all four sides is defined on the parent Group block. It's generally pretty easy to create. 

With this PR though, this pattern looks like this by default: 

Screen Shot 2021-12-13 at 1 36 06 PM

I can't override that `0` top padding on the Group block since I can't override padding for a block that needs default/wide-width children. Instead, I have to create _another_ group block to house the children, and add top/bottom margin to that: 

layout

Most users won't figure out how to make that happen — it's far less intuitive than just editing the padding for the Gray Group block.

**Comment by jasmussen** [comment]
It feels worth adding that for TT2 we did discuss whether we couldn't handle the challenge in [CSS](https://github.com/WordPress/twentytwentytwo/pull/289#issuecomment-990975662). It has its own tradeoff, but it could buy some time to figure out a better solution, potentially.

**Comment by youknowriad** [comment]
The CSS solutions that I saw so far will solve some use-cases but create issues for others, I don't believe there's a perfect CSS solution for this.

What makes this hard is that ideally, there shouldn't be any user or theme author interactions, the premise is simple:

Applying a full width alignment to a block should ignore the padding of its container, the reality is that we can't write a CSS rule to achieve that in a generic way.

**Comment by kjellr** [comment]
> What makes this hard is that ideally, there shouldn't be any user or theme author interactions, the premise is simple:
> 
> Applying a full width alignment to a block should ignore the padding of its container, the reality is that we can't write a CSS rule to achieve that in a generic way.

That's a good summary! This is truly difficult. 😅

At the moment, I'm primarily trying to consider the best-case scenario for 5.9... Since we don't have a solution we all feel solid about, I think anything we land at this stage needs to be relatively minimal and mostly harmless: Something automatic, that can be stripped out and replaced if need be. It would be good to get something into today's Beta, since it's the last (confirmed) beta before the holidays.

I lean towards the fix in https://github.com/WordPress/twentytwentytwo/pull/291 because it _mostly_ works, and is just a handful of CSS rules. I think the main thing going for it is is that it's replaceable without setting precedent, and it buys us a bit more time. For now, I think we should just try putting that in and seeing how it tests during the next beta cycle.

**Comment by youknowriad** [comment]
> I lean towards the fix in WordPress/twentytwentytwo#291 because it mostly works, and is just a handful of CSS rules. I think the main thing going for it is is that it's replaceable without setting precedent, and it buys us a bit more time. For now, I think we should just try putting that in and seeing how it tests during the next beta cycle.

I can be onboard with that but let's keep it to the real CSS minimum override and also add a clarifying comment that block themes shouldn't rely on it too much as it might break or become unnecessary in the future. Themes will consider 2022 as the theme to copy for everything.

**Comment by kjellr** [comment]
> I can be onboard with that but let's keep it to the real CSS minimum override and also add a clarifying comment that block themes shouldn't rely on it too much as it might break or become unnecessary in the future. Themes will consider 2022 as the theme to copy for everything.

Fully agreed. 👍

**Comment by MaggieCabrera** [comment]
> Themes will consider 2022 as the theme to copy for everything.

That's why it's important that the fix is in the editor, so future themes don't include these extra rules that will potentially be obsolete in the near future in favor of a better solution. I think the TT2 styles on https://github.com/WordPress/twentytwentytwo/pull/291 with the variable for controlling the padding amount via theme.json should be a good starting point to make it opt-in.

**Comment by glendaviesnz** [comment]
FYI - the number of changes in layout, etc. made a rebase a bit messy, so have cherry-picked the changes onto a fresh branch off trunk [over here](https://github.com/WordPress/gutenberg/pull/39871) so we can do some more exploration with this.

**Comment by youknowriad** [comment]
Interesting @glendaviesnz I'm closing this now
